packages/react-graphql/src/hooks/useGraphQLDocument.ts

```diff
@@ -0,0 +1,48 @@
+import {useState, useEffect} from 'react';
+import {OperationVariables} from 'apollo-client';
+import {DocumentNode} from 'graphql-typed';
+
+import {AsyncQueryComponentType} from '..';
+
+export default function useGraphQLDocument<
+  Data = any,
+  Variables = OperationVariables,
+  DeepPartial = {}
+>(
+  documentOrComponent:
+    | DocumentNode<Data, Variables>
+    | AsyncQueryComponentType<Data, Variables, DeepPartial>,
+) {
+  const [document, setDocument] = useState<DocumentNode<
+    Data,
+    Variables
+  > | null>(
+    () => (isDocumentNode(documentOrComponent) ? documentOrComponent : null),
+  );
+
+  async function loadDocument() {
+    if (!isDocumentNode(documentOrComponent)) {
+      try {
+        const resolved = await documentOrComponent.resolve();
+        setDocument(resolved);
+      } catch (error) {
+        throw Error('error loading GraphQL document');
+      }
+    }
+  }
+
+  useEffect(
+    () => {
+      if (document === null) {
+        loadDocument();
+      }
+    },
+    [documentOrComponent],
```

Comment:
`setDocument` is a setter so assuming that won't change.
I could include `document` but I also don't see a scenario where I would want to reload the document because `query` changed.

Reply:
> setDocument is a setter so assuming that won't change

True, however, I was under the assumption is it a good practice all dependencies in cases like these. c/c @lemonmade 

Same goes for `loadDocument`. Thoughts?

I can see though how we can think of this as `componentDidMount` where we want to run this effect only on a particular change (which, in this case, is `documentOrComponent` )

Reply:
The `setter` function from `useState` is not required no.